src/fluree/db/util/validation.cljc

```diff
@@ -0,0 +1,73 @@
+(ns fluree.db.util.validation
+  (:require [clojure.string :as str]
+            [malli.core :as m]
+            [fluree.db.util.log :as log]))
+
+(def non-empty-string
+  (m/schema [:string {:min 1}]))
+
+(def value? (complement coll?))
+
+(defn qualified-keyword->json-ld
+  [kw]
+  (str (namespace kw) ":" (name kw)))
+
+(defn compact-iri->keyword
+  "Converts compact IRI strings to keywords. If there is a colon in the iri,
+  the part before the colon becomes the keyword's namespace.
+  E.g.
+  \"foo\" -> :foo
+  \"foo:bar\" -> :foo/bar"
+  [iri]
+  (-> iri
+      (str/split #":")
+      (->> (cons nil)
+           (take-last 2)
+           (apply keyword))))
+
+(def iri-key
+  "Decodes all string values to keywords even if they don't look like compact
+  IRIs. Intended to support e.g. \"id\" -> :id."
+  (m/schema
+    [:orn
+     {:decode/json
+      (fn [v]
+        (log/debug "decoding iri key:" v)
+        (if (string? v)
+          (if (str/includes? v "://") ; non-compact IRI
```

Comment:
I think this non-compact iri test may need to be a bit more sophisticated - here's the spec: https://www.rfc-editor.org/rfc/rfc3987#section-2

Reply:
Are there some that could contain a colon but not contain `://`? That's all this needs to ensure.

Reply:
Yeah, I think a colon is all you need to be an iri: `foo:foo` is a valid fully expanded iri.

Reply:
Which should or should not be converted into `:foo/foo`?

Reply:
None of these should get touched. We’d never want to convert id => :id nor convert a string compact iri to a keyword compact iri as the @context meaning would break. 

This would not only produce incorrect results but the concept of it introduces a bunch of extra compute for no real value, even if it would work (which it won’t). 

Maybe I’m misunderstanding how this is intended to be used, is there a ticket for this?

Reply:
Re: a colon designating an iri - we can/will store relative iris which would not have a colon until expanded with a default @base context.